# This is a patch against Gtk+ 2.0.5 sources that fixes Windows specific
# problems.

###############################################################################
#
###############################################################################
# Problems already included in the Gtk+ CVS:


###############################################################################
#
###############################################################################
# Problems not included in CVS:



#########################################
# Ctrl-C
#########################################

--- gtk+-2.0.5/gdk/win32/save/gdkevents-win32.c      Tue Jun  4 19:31:17 2002
+++ gtk+-2.0.5/gdk/win32/gdkevents-win32.c   Wed Jun  5 11:08:12 2002
@@ -886,6 +886,11 @@
 static guint
 vk_from_char (guint c)
 {
+  if ((c >= 1) && (c <=26))
+    {
+      return (c - 1 + 'A') ;
+    }
   switch (c)
     {
     case '\b':


#########################################
# always on top
#########################################

--- gtk+-2.0.5/gdk/win32/save/gdkwindow-win32.c      Fri May 31 12:18:44 2002
+++ gtk+-2.0.5/gdk/win32/gdkwindow-win32.c   Wed Jun  5 15:51:34 2002
@@ -1519,7 +1516,9 @@
   if (!SetParent (window_id, parent_id))
        WIN32_API_FAILED ("SetParent");
 #else /* make the modal window topmost instead */
-  if (!SetWindowPos (window_id, HWND_TOPMOST, 0, 0, 0, 0,
+  if (!SetWindowPos (window_id, HWND_NOTOPMOST, 0, 0, 0, 0,
                      SWP_NOMOVE | SWP_NOSIZE))
     WIN32_API_FAILED ("SetWindowPos");
 #endif



#########################################
# boxes display (e.g. tab characters)
#########################################

--- pango-1.0.3/pango/save/pangowin32.c   Thu Jun  6 11:27:34 2002
+++ pango-1.0.3/pango/pangowin32.c        Thu Jun  6 14:07:58 2002
@@ -240,7 +240,7 @@
 {
   HFONT old_hfont = NULL;
   HFONT hfont;
-  int i;
+  int i, idx ;
   guint16 *glyph_indexes;
   INT *dX;

@@ -255,16 +255,23 @@
   glyph_indexes = g_new (guint16, glyphs->num_glyphs);
   dX = g_new (INT, glyphs->num_glyphs);

+  idx = 0 ;
   for (i = 0; i <glyphs->num_glyphs; i++)
     {
-      glyph_indexes[i] = glyphs->glyphs[i].glyph;
-      dX[i] = glyphs->glyphs[i].geometry.width / PANGO_SCALE;
+      /* fduguet : skips the zero glyphs */
+      if (glyphs->glyphs[i].glyph != 0)
+        {
+          /* glyph is handeled */
+          glyph_indexes[idx] = glyphs->glyphs[i].glyph ;
+          dX[idx] = glyphs->glyphs[i].geometry.width / PANGO_SCALE ;
+          ++idx ;
+        }
     }

   ExtTextOutW (hdc, x, y,
               ETO_GLYPH_INDEX,
               NULL,
-              glyph_indexes, glyphs->num_glyphs,
+              glyph_indexes, idx,
               dX);

   SelectObject (hdc, old_hfont); /* restore */


#########################################
# memory leak - GdkImage not unreferenced
#########################################

--- gtk+-2.0.5/gdk/win32/save/gdkpixmap-win32.c Fri Jun 21 16:08:55 2002
+++ gtk+-2.0.5/gdk/win32/gdkpixmap-win32.c Fri Jun 21 16:10:25 2002
@@ -101,6 +101,7 @@
 {
   GdkPixmapImplWin32 *impl = GDK_PIXMAP_IMPL_WIN32 (object);
   GdkPixmap *wrapper = GDK_PIXMAP (GDK_DRAWABLE_IMPL_WIN32 (impl)->wrapper);
+  GdkImage *image = impl->image ;
 
   GDK_NOTE (PIXMAP, g_print ("gdk_pixmap_impl_win32_finalize: %p\n",
 			     GDK_PIXMAP_HBITMAP (wrapper)));
@@ -109,6 +110,9 @@
     WIN32_GDI_FAILED ("DeleteObject");
 
   gdk_win32_handle_table_remove (GDK_PIXMAP_HBITMAP (wrapper));
+
+  image->windowing_data = NULL ; /* avoiding loop in unref */
+  g_object_unref (image) ;
 
   G_OBJECT_CLASS (parent_class)->finalize (object);
 }


#########################################
# memory leak - increasing size of translate_queue
#########################################

--- gtk+-2.0.5/gdk/win32/save/gdkgeometry-win32.c Fri Jun 21 15:58:00 2002
+++ gtk+-2.0.5/gdk/win32/gdkgeometry-win32.c Fri Jun 21 16:29:24 2002
@@ -769,6 +769,25 @@
     }
 }
 
+
+static void
+gdk_window_queue_append (GdkWindow *window, GdkWindowQueueItem *item)
+{
+  int i ;
+
+  translate_queue = g_slist_append (translate_queue, item) ;
+
+  if (g_slist_length (translate_queue) >= 128)
+    {
+      GdkWindowImplWin32 *impl = GDK_WINDOW_IMPL_WIN32 (GDK_WINDOW_OBJECT (window)->impl) ;
+      GdkRegion *whole_region = gdk_region_rectangle (&impl->position_info.clip_rect) ;
+      _gdk_window_process_expose (window,
+                          GetCurrentTime (),
+                          whole_region) ;
+      gdk_region_destroy (whole_region) ;
+    }
+}
+
 static void
 gdk_window_queue_translation (GdkWindow *window,
 			      gint       dx,
@@ -787,7 +806,7 @@
 			     dx, dy));
 
   gdk_drawable_ref (window);
-  translate_queue = g_slist_append (translate_queue, item);
+  gdk_window_queue_append (window, item) ;
 }
 
 gboolean
@@ -810,7 +829,7 @@
 			     area->extents.x1, area->extents.y1));
 
   gdk_drawable_ref (window);
-  translate_queue = g_slist_append (translate_queue, item);
+  gdk_window_queue_append (window, item) ;
 
   return TRUE;
 #else



#########################################
# linefeeds in copy-paste
#########################################

--- gtk+-2.0.5/gdk/win32/save/gdkproperty-win32.c Fri Jun 21 18:14:52 2002
+++ gtk+-2.0.5/gdk/win32/gdkproperty-win32.c Fri Jun 21 18:14:52 2002
@@ -305,7 +305,7 @@
 	  method = PLAIN_ASCII;
 	  size = nelements;
 	  for (i = 0; i < nelements; i++)
-	    if (*data == '\n')
+	    if (data[i] == '\n')
 	      size++;
 	  size++;
 	  GDK_NOTE (DND, g_print ("...as text: %.40s\n", data));


#########################################
# Selection problem B430-008
#########################################


--- save/gtktextview.c	Wed Jul  3 12:42:28 2002
+++ gtktextview.c	Wed Jul  3 11:52:51 2002
@@ -3866,8 +3866,11 @@
 
   text_view = GTK_TEXT_VIEW (widget);
 
+  /* 
   if (event->window != text_view->text_window->bin_window)
     return FALSE;
+  */
 
   if (event->button == 1)
     {
